File: 207-course-schedule/207-course-schedule.cpp
class Solution {
    map<int, list<int>> adj;
    unordered_set<int> visited;
public:
    bool canFinish(int nc, vector<vector<int>>& prereq) {
        // creating adj list
        for (auto &lists : prereq) {
            adj[lists[0]].push_back(lists[1]);
        }
        // CHECK for dis-connected component
        for (int i = 0;i < nc;i++) {
            if (visited.find(i) == visited.end()) {
                if (!dfs(i)) 
                    return false;
            }
        }
        return true;
    }
private:
    bool dfs(int sv) {
        // check for cycles
        if (visited.find(sv) != visited.end()) 
            return false;
        if (adj[sv].empty()) 
            return true;
        visited.insert(sv);
        for (auto &node : adj[sv]) {
            if (!dfs(node)) 
                return false;
        }
        // here in this point we if we don't find any loop then we reached a position where that course could be complete so we can remove that dependency
        adj[sv].clear();
        visited.erase(sv);
        return true;
    }
};

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, n: int, pre: List[List[int]]) -> bool:
        adj = {i: [] for i in range(n+1)}
        for si, ei in pre:
            adj[si].append(ei)
        rec, vis = set(), set()
        def isCycle(node):
            vis.add(node)
            rec.add(node)
            for nd in adj[node]:
                if not nd in vis:
                    if isCycle(nd):
                        return True
                elif nd in rec:
                    return True
            rec.remove(node)
            return False
        for i in range(n):
            if isCycle(i):
                return False
        return True
        